Support Tensorflow Lite for Windows

mattn opened the pull request:
This change is not fully porting of Tensorflow Lite since some fixes for dependency projects are required on Windows. ex: gemmlowp, farmhash. However I could build libtensorflow-lite.a using this patch and some small fixes to the dependencies.

https://github.com/mattn/webcam-detect-tflite

This is example app to detect objects on real time camera streaming using TensorFlow Lite on Windows.

![image](https://user-images.githubusercontent.com/10111/53289235-e2da5800-37d6-11e9-9bfc-2db20b47e5aa.png)

mattn committed:
Support Tensorflow Lite for Windows

mattn committed:
Support Tensorflow Lite for Windows

mattn commented:
This is a patch for gemmlowp and farmhash.

```diff
diff -ur downloads.orig/farmhash/src/farmhash.cc downloads/farmhash/src/farmhash.cc
--- downloads.orig/farmhash/src/farmhash.cc	2017-09-14 05:13:37.000000000 +0900
+++ downloads/farmhash/src/farmhash.cc	2019-02-24 01:17:52.840988000 +0900
@@ -114,7 +114,7 @@
 
 #if defined(FARMHASH_UNKNOWN_ENDIAN) || !defined(bswap_64)
 
-#ifdef _MSC_VER
+#ifdef _WIN32
 
 #undef bswap_32
 #undef bswap_64
diff -ur downloads.orig/gemmlowp/internal/platform.h downloads/gemmlowp/internal/platform.h
--- downloads.orig/gemmlowp/internal/platform.h	2019-02-24 02:14:49.938967500 +0900
+++ downloads/gemmlowp/internal/platform.h	2019-02-24 00:53:31.840022600 +0900
@@ -69,12 +69,21 @@
   return max_threads;
 }
 
+#ifdef __MINGW64__
+inline double real_time_in_seconds() {
+  __int64 wintime;
+  GetSystemTimeAsFileTime((FILETIME *)&wintime);
+  wintime -= 116444736000000000;  // 1jan1601 to 1jan1970
+  return wintime / 10000000 + wintime % 10000000 * 100 * 1e-9;
+}
+#else
 inline double real_time_in_seconds() {
   __int64 wintime;
   GetSystemTimeAsFileTime((FILETIME *)&wintime);
   wintime -= 116444736000000000i64;  // 1jan1601 to 1jan1970
   return wintime / 10000000i64 + wintime % 10000000i64 * 100 * 1e-9;
 }
+#endif
 
 #else
 inline void *aligned_alloc(size_t alignment, size_t size) {
```

shahzadlone commented on tensorflow/lite/mmap_allocation.cc:
```diff
@@ -24,6 +28,52 @@ limitations under the License.
+  buffer_size_bytes_ = GetFileSize( file_handle_, nullptr );
+
+  file_mapping_ = CreateFileMapping(file_handle_, NULL, PAGE_READONLY, 0, 0, NULL);
+  if (file_mapping_ == NULL)
```
```nullptr```?

mattn committed:
Use nullptr

mattn committed:
Use nullptr

mattn committed:
Cosmetic changes

mattn committed:
Cosmetic changes

mattn committed:
Merge branch 'fix-tflite-build' into tflite-windows

mattn committed:
Merge branch 'tflite-windows' of github.com:mattn/tensorflow into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Add missing memmem

mattn committed:
Add missing memmem

mattn committed:
Merge branch 'tflite-windows' of github.com:mattn/tensorflow into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

jdduke commented:
What compiler+version are you using? We've been able to successfully build with Visual Studio for some time now, though I believe only with more recent versions.

mattn commented:
I use mingw64 compiler (which is bundled on msys2/mingw64)

```
gcc (Rev2, Built by MSYS2 project) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

mattn commented:
The changes for `mmap_allocation.cc` is not required. So I'll remove  this.

mattn committed:
Merge branch 'master' into tflite-windows

jdduke commented:
Taking a step back, I'm wondering if we have a broader problem. Should the [//third_party/tensorflow:windows condition](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/BUILD#L167) be satisfied when building with MinGW? Does TensorFlow support building with MinGW? I'd prefer to fix this at the build rule level, if we can, rather than sprinkling more #ifdefs int he code.

mattn commented:
TensorFlow does not support MinGW. https://github.com/tensorflow/tensorflow/issues/22840#issuecomment-429502287

But mingw can build tensorflow-lite using tensorflow/lite/tools/make/Makefile.

jdduke commented on tensorflow/lite/tools/make/Makefile:
```diff
@@ -128,6 +132,9 @@ BUILD_WITH_NNAPI=true
 ifeq ($(BUILD_TYPE),micro)
```
Let's also make the windows change for mmap_allocation.cc vs mmap_allocation_disabled.cc above, and then let's undo the changes to mmap_allocation.cc and nnapi_delegate.cc

jdduke commented on tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc:
```diff
@@ -49,6 +49,29 @@ void* aligned_alloc(size_t alignment, size_t size, void** freeing_buffer) {
+
 // Use /proc/cpuinfo to test whether we have the right processor.
 bool HasSdotInstruction() {
   // TODO(strohman): Replace this with a proper API call once we are running
```
Rather than adding memem just for this, let's just add the #ifdef here for __MINGW32__ and always return false.

jdduke commented on tensorflow/lite/allocation.h:
```diff
@@ -24,6 +24,9 @@ limitations under the License.
 #include "tensorflow/lite/simple_memory_arena.h"
 #include "tensorflow/lite/string.h"
+#ifdef _WIN32
+#include <windows.h>
```
We shouldn't need these changes if we switch to using mmap_allocation_disabled.cc. See comment below.

mattn committed:
Merge branch 'master' into tflite-windows

mattn commented on tensorflow/lite/tools/make/Makefile:
```diff
@@ -154,7 +161,9 @@ TF_LITE_CC_SRCS := $(filter-out $(CORE_CC_EXCLUDE_SRCS), $(CORE_CC_ALL_SRCS))
 BENCHMARK_SRCS_DIR := tensorflow/lite/tools/benchmark
-BENCHMARK_ALL_SRCS := $(TFLITE_CC_SRCS) \
+BENCHMARK_ALL_SRCS := $(TF_LITE_CC_SRCS) \
+	tensorflow/lite/testing/init_tensorflow.cc \
```
Because the symbol `tflite::InitTensorFlow()` is not inculded the benchmark-lib.a

mattn committed:
Fixes about review points

mattn committed:
To build on Windows, this file should not be symbolic link.

mattn commented on tensorflow/lite/nnapi_delegate.cc:
```diff
@@ -15,7 +15,6 @@ limitations under the License.
 
 #include "tensorflow/lite/nnapi_delegate.h"
 #include <fcntl.h>
-#include <sys/mman.h>
```
sorry, will revert this delete soon.

mattn commented on tensorflow/lite/tools/make/Makefile:
```diff
@@ -154,7 +168,8 @@ TF_LITE_CC_SRCS := $(filter-out $(CORE_CC_EXCLUDE_SRCS), $(CORE_CC_ALL_SRCS))
 BENCHMARK_SRCS_DIR := tensorflow/lite/tools/benchmark
-BENCHMARK_ALL_SRCS := $(TFLITE_CC_SRCS) \
+BENCHMARK_ALL_SRCS := $(TF_LITE_CC_SRCS) \
+	$(wildcard $(TESTING_SRCS_DIR)/*.cc) \
```
will remove this too.

mattn committed:
Revert some changes

mattn commented:
@jdduke Fixed as you pointed. Thanks. Still not possible to build benchmark.

```
g++ -O3 -DNDEBUG -fPIC  --std=c++11 -fext-numeric-literals -D__LITTLE_ENDIAN__ -I. -Ic:/dev/tensorflow/tensorflow/lite/tools/make/../../../../../ -Ic:/dev/tensorflow/tensorflow/lite/tools/make/../../../../../../ -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/ -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/eigen -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/absl -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/gemmlowp -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/neon_2_sse -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/farmhash/src -Ic:/dev/tensorflow/tensorflow/lite/tools/make/downloads/flatbuffers/include -I -I/usr/local/include \
-o c:/dev/tensorflow/tensorflow/lite/tools/make/gen/windows_x86_64/bin/benchmark_model \
 c:/dev/tensorflow/tensorflow/lite/tools/make/gen/windows_x86_64/lib/benchmark-lib.a  -lstdc++ -lpthread -lm -lz
c:/dev/tensorflow/tensorflow/lite/tools/make/gen/windows_x86_64/lib/benchmark-lib.a(benchmark_plus_flex_main.o):benchmark_plus_flex_main.cc:(.text+0x3b): undefined reference to `tflite::InitTensorFlow()'
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [tensorflow\lite\tools\make\Makefile:254: c:/dev/tensorflow/tensorflow/lite/tools/make/gen/windows_x86_64/bin/benchmark_model] Error 1
```

jdduke commented:
We should exclude benchmark_plus_flex_main.cc from `BENCHMARK_SRCS`.

mattn committed:
Merge branch 'tflite-windows' of github.com:mattn/tensorflow into tflite-windows

mattn committed:
Exclude benchmark_plus_flex_main.cc from BENCHMARK_SRCS

mattn committed:
Merge branch 'master' into tflite-windows

jdduke commented on tensorflow/lite/kernels/internal/optimized/neon_tensor_utils.cc:
```diff
@@ -90,6 +93,7 @@ bool HasSdotInstruction() {
     return found;
   }();
   return has_sdot;
+#endif
```
#endif  // __MINGW32__

jdduke commented on tensorflow/lite/experimental/c/c_api_types.h:
```diff
@@ -1 +1 @@
-../../c/c_api_internal.h
\ No newline at end of file
+#include "../../c/c_api_internal.h"
```
I'm not sure this will work, is it 100% necessary to build? Can you exclude it from the srcs?

mattn replied:
Yes, do you mean that need to change following files?

```
experimental/c/BUILD|43 col 10| "c_api_types.h",
experimental/c/BUILD|61 col 10| "c_api_types.h",
experimental/c/c_api.cc|21 col 42| #include "tensorflow/lite/experimental/c/c_api_types.h"
experimental/c/c_api.h|24 col 11| #include "c_api_types.h"
experimental/ios/BUILD.apple|24 col 43| "//tensorflow/lite/experimental/c:c_api_types.h",
experimental/objc/TensorFlowLiteObjC.podspec|31 col 31| tfl_dir + 'experimental/c/c_api_types.h',
```

jdduke replied:
Hmm, I need to think a bit more about this one, as it will likely fail during our internal submit.

If you don't mind, can we drop the change to this file, and keep the other changes? I can approve that, and get back to you about a fix for this file.

jdduke reviewed (changes requested):
Almost there, thanks for your patience.

mattn committed:
Cosmetic change

mattn committed:
Revert change for c_api_types.h

mattn commented:
Reverted change for c_api_types.h

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'master' into tflite-windows

mattn committed:
Merge branch 'tflite-windows' of github.com:mattn/tensorflow into tflite-windows

rthadur commented:
@mattn can you please resolve conflicts

mattn committed:
Merge branch 'master' into tflite-windows

mattn commented:
Fixed conflict. Also updated patch for third-party libraries.

```diff
diff -ur downloads.orig/farmhash/src/farmhash.cc downloads/farmhash/src/farmhash.cc
--- downloads.orig/farmhash/src/farmhash.cc	2019-05-05 22:39:31.645761900 +0900
+++ downloads/farmhash/src/farmhash.cc	2019-05-05 22:46:58.122055200 +0900
@@ -114,7 +114,7 @@
 
 #if defined(FARMHASH_UNKNOWN_ENDIAN) || !defined(bswap_64)
 
-#ifdef _MSC_VER
+#ifdef _WIN32
 
 #undef bswap_32
 #undef bswap_64
@@ -216,7 +216,7 @@
   return shift == 0 ? val : ((val >> shift) | (val << (64 - shift)));
 }
 
-#if defined(_MSC_VER) && defined(FARMHASH_ROTR)
+#if defined(_WIN32) && defined(FARMHASH_ROTR)
 
 STATIC_INLINE uint32_t Rotate32(uint32_t val, int shift) {
   return sizeof(unsigned long) == sizeof(val) ?
```

mattn commented:
FYI, I sent small fix to farmhash https://github.com/google/farmhash/pull/32

So we will be possible to build tensorflow-lite with mingw compiler. (without patches)

mattn commented:
https://github.com/google/farmhash/pull/32 was merged.
